Xamarin.Forms.Core/Profiler.cs

```diff
@@ -0,0 +1,123 @@
+﻿using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Diagnostics;
+using System.IO;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Xamarin.Forms.Internals;
+
+namespace Xamarin.Forms.Internal
+{
+	[EditorBrowsable(EditorBrowsableState.Never)]
+	public struct Profile : IDisposable
+	{
+		const int Capacity = 1000;
+		const string PROFILE = "PROFILE";
+		public struct Datum
+		{
+			public string Name;
+			public string Id;
+			public long Ticks;
+			public int Depth;
+			public int Line;
+		}
+		public static List<Datum> Data = new List<Datum>(Capacity);
+
+		static Stack<Profile> s_stack = new Stack<Profile>(Capacity);
+		static int s_depth = 0;
+		static bool s_stopped = false;
+		static Stopwatch s_stopwatch = new Stopwatch();
+
+		readonly long start;
+		readonly string name;
+		readonly int slot;
+		readonly string id;
+		readonly int line;
+
+		public static void Stop()
+		{
+			// unwind stack
+			s_stopped = true;
+			while (s_stack.Count > 0)
+				s_stack.Pop();
+		}
+		[Conditional(PROFILE)]
+		public static void Push(
```

Comment:
I like the way this is designed, but I'm wondering about calling the methods `Push`, `Pop`, and `PopPush`. That sort of requires that the user knowing that the profiler is a stack, internally. If I want to start profiling something, I push it onto the stack, if I want to stop profiling it I pop it off the stack - would it be easier for a future user if the methods were just called `Start`, `Stop`, and something like `Checkpoint` for `PopPush`? 

Not a deal breaker, since everyone using this will be be, by definition, a programmer who should know what a stack is. But it might be easier for new folks to pick up (and for use to read when reviewing).

Reply:
Oh, well, sure. If it's `start` then there will be many nested `start` calls which seems odd.. Sucker is taking samples inside stack frames so `push` and `pop` seemed reasonable. I could get behind changing `PopPush` to `Sample`. How about that?

Reply:
I can live with `PopPush` -> `Sample` and leaving the others as-is.

Reply:
How about `Start` and `Stop` to enable the profiler. Without calling `Start` all the public entry points are noops and are small enough so they should be in-lined so that their cost is minimal and we're able to simply ship the profiler disabled in the product.

Then `Push`, `Pop` and `PopPush` become `SampleStart`, `SampleEnd` and `SamplePartition`.